### DISCUSSION TITLE
Is it possible to specify image url to be included in <link rel="prelod" .../> in <head>?

### DISCUSSION THREAD

**Discussion by acromatica** [breezefront/community#332] [1. Help]
Hi. On the home page I'm building mobile and desktop versions have different hero images. In the console I'm getting a warning that mobile image was preloaded, but not used, which makes sense since I'm checking the site on desktop screen.

HTML code of the image is this:
`<picture>
	<source media="(min-width: 768px)" type="image/webp" srcset="hero-desk@1x.webp, hero-desk@2x.webp 2x, hero-desk@3x.webp 3x">
	<source media="(min-width: 768px)" type="image/jpeg" srcset="hero-desk@1x.jpg, hero-desk@2x.jpg 2x, hero-desk@3x.jpg 3x">
	<source type="image/webp" srcset="hero-mob@1x.webp, hero-mob@2x.webp 2x, hero-mob@3x.webp 3x">
	<source type="image/jpeg" srcset="hero-mob@1x.jpg, hero-mob@2x.jpg 2x, hero-mob@3x.jpg 3x">
	<img src="hero-mob@1x.jpg" srcset="hero-mob@1x.jpg, hero-mob@2x.jpg 2x, hero-mob@3x.jpg 3x" alt="">
</picture>`

Basically one image for mobile and one image for desktop in @1x, @2x and @3x resolutions, with a fallback to jpg format.

Generated preload link looks like this:
`<link rel="preload" fetchpriority="high" as="image" href="hero-mobbz-at1x.jpg" imagesrcset="hero-mobbz-at1x.jpg, hero-mobbz-at2x.jpg 2x, hero-mobbz-at3x.jpg 3x">`

The `bz-at` part is probably @ symbol which I can replace with something else, but is there a way to update the rest? Or remove this preload link altogether, since the output doesn't really make sense.

In this case it should be:

`<link rel="preload" imagesrcset="hero-mob@1x.webp 1x, hero-mob@2x.webp 2x, hero-mob@3x.webp 3x" as="image" media="(max-width: 768px)">`
`<link rel="preload" imagesrcset="hero-desk@1x.webp 1x, hero-desk@2x.webp 2x, hero-desk@3x.webp 3x" as="image" media="(min-width: 768.1px)">`

Thank you!

**Comment by vovayatsyuk** [comment]
Thanks for reporting. It's a bug in DOM processing. I just fixed it in https://github.com/breezefront/module-breeze/commit/e6373a80bbefcee4254dac123f24952c2c07aacb

Please note, that it's not released yet.

**Comment by acromatica** [comment]
Thanks for looking into this @vovayatsyuk! 
I've added the changes from the commit to see if they fixed the issue; it's partly resolved, now filename contains @ instead of bz-at, but the rest remained the same: the browser preloads incorrect image.

Here's the preload link that I see on desktop:
`<link rel="preload" fetchpriority="high" as="image" href="hero-mob@1x.jpg" imagesrcset="hero-mob@1x.jpg, hero-mob@2x.jpg 2x,hero-mob@3x.jpg 3x">`

Preloaded file is mobile .jpg (fallback for .webp), while I'm opening the page on a desktop screen in a browser that supports webp format. In order to preload correct images, the code should be the following:

`<link rel="preload" imagesrcset="hero-mob@1x.webp 1x, hero-mob@2x.webp 2x, hero-mob@3x.webp 3x" as="image" media="(max-width: 768px)">`
`<link rel="preload" imagesrcset="hero-desk@1x.webp 1x, hero-desk@2x.webp 2x, hero-desk@3x.webp 3x" as="image" media="(min-width: 768.1px)">`

With this, when the page is opened on mobile, the first link will preload mobile .webp image, and if it's a desktop screen the second link will preload desktop .webp image. In both cases it doesn't make sense to be preloading a fallback .jpg because if the browser understands `<link rel="preload" .../>` it supports webp images too.
So, is there a way to remove `<link rel="preload" .../>` for now if it can't be adjusted to preload the correct images?
Thank you.

**Reply by vovayatsyuk** [reply]
Yeah, that is pretty tricky. I can propose the following solution:

1. Do not preload an image from `<picture>` if more than one source is available.
2. But preload the image from `<picture>` if the source has the `data-preload` attribute.

Example of your HTML to get images with preload:

```html
<picture> 
    <source data-preload media="(min-width: 768px)" type="image/webp" srcset="hero-desk@1x.webp, hero-desk@2x.webp 2x, hero-desk@3x.webp 3x">
    <source media="(min-width: 768px)" type="image/jpeg" srcset="hero-desk@1x.jpg, hero-desk@2x.jpg 2x, hero-desk@3x.jpg 3x">
    <source data-preload type="image/webp" srcset="hero-mob@1x.webp, hero-mob@2x.webp 2x, hero-mob@3x.webp 3x">
    <source type="image/jpeg" srcset="hero-mob@1x.jpg, hero-mob@2x.jpg 2x, hero-mob@3x.jpg 3x">
    <img src="hero-mob@1x.jpg" srcset="hero-mob@1x.jpg, hero-mob@2x.jpg 2x, hero-mob@3x.jpg 3x" alt="">
</picture>
```

Will that work for you?

**Reply by acromatica** [reply]
`data-preload` sounds great! 
And I guess in future it could be modified to work like solution #1, something like `data-preload="false"` on the <img ../> or <picture ../> element not to preload anything.

**Reply by vovayatsyuk** [reply]
1. We will not preload the `<img>` inside the picture in the next breeze version: https://github.com/breezefront/module-breeze/commit/edf7a30068fa2ca7dc9c59024417c0d95d4a08e1
2. Added the ability to manually preload images using the `data-preload` **CSS class name**. I ended up with a class name because Magento's Pagebuilder allows editing it: https://github.com/breezefront/module-breeze/commit/2d11eb5922a41d80734ff9f6526a062a950ea248
3. Sources are not supported yet.

**Reply by acromatica** [reply]
Cool, thank you so much for the update. Looking forward to sources update!
